README.md
oauth2-server-php
=================

[![Build Status](https://travis-ci.org/bshaffer/oauth2-server-php.svg?branch=develop)](https://travis-ci.org/bshaffer/oauth2-server-php)

[![Total Downloads](https://poser.pugx.org/bshaffer/oauth2-server-php/downloads.png)](https://packagist.org/packages/bshaffer/oauth2-server-php)

[![Coverage Status](https://coveralls.io/repos/bshaffer/oauth2-server-php/badge.svg?branch=develop&service=github)](https://coveralls.io/github/bshaffer/oauth2-server-php?branch=develop)

View the [complete documentation](http://bshaffer.github.io/oauth2-server-php-docs/)


CHANGELOG.md
CHANGELOG for 1.x
=================

This changelog references the relevant changes (bug and security fixes) done
in 1.x minor versions.

To see the files changed for a given bug, go to https://github.com/bshaffer/oauth2-server-php/issues/### where ### is the bug number
To get the diff between two versions, go to https://github.com/bshaffer/oauth2-server-php/compare/v1.0...v1.1
To get the diff for a specific change, go to https://github.com/bshaffer/oauth2-server-php/commit/XXX where XXX is the change hash

* 1.9.0 (2016-01-06)

  PR: https://github.com/bshaffer/oauth2-server-php/pull/788

  * bug #645 - Allow null for client_secret
  * bug #651 - Fix bug in isPublicClient of Cassandra Storage
  * bug #670 - Bug in client's scope restriction
  * bug #672 - Implemented method to override the password hashing algorithm
  * bug #698 - Fix Token Response's Content-Type to application/json
  * bug #729 - Ensures unsetAccessToken and unsetRefreshToken return a bool
  * bug #749 - Fix UserClaims for CodeIdToken
  * bug #784 - RFC6750 compatibility
  * bug #776 - Fix "redirect_uri_mismatch" for URIs with encoded characters
  * bug #759 - no access token supplied to resource controller results in empty request body
  * bug #773 - Use OpenSSL random method before attempting Mcrypt's.
  * bug #790 - Add mongo db

* 1.8.0 (2015-09-18)

  PR: https://github.com/bshaffer/oauth2-server-php/pull/643

  * bug #594 - adds jti
  * bug #598 - fixes lifetime configurations for JWTs
  * bug #634 - fixes travis builds, upgrade to containers
  * bug #586 - support for revoking tokens
  * bug #636 - Adds FirebaseJWT bridge
  * bug #639 - Mongo HHVM compatibility

* 1.7.0 (2015-04-23)

  PR: https://github.com/bshaffer/oauth2-server-php/pull/572

  * bug #500 - PDO fetch mode changed from FETCH_BOTH to FETCH_ASSOC
  * bug #508 - Case insensitive for Bearer token header name  ba716d4
  * bug #512 - validateRedirectUri is now public
  * bug #530 - Add PublicKeyInterface, UserClaimsInterface to Cassandra Storage
  * bug #505 - DynamoDB storage fixes
  * bug #556 - adds "code id_token" return type to openid connect
  * bug #563 - Include "issuer" config key for JwtAccessToken
  * bug #564 - Fixes JWT vulnerability
  * bug #571 - Added unset_refresh_token_after_use option

* 1.6 (2015-01-16)

  PR: https://github.com/bshaffer/oauth2-server-php/pull/496

  * bug 437 - renames CryptoToken to JwtAccessToken / use_crypto_tokens to use_jwt_access_tokens
  * bug 447 - Adds a Couchbase storage implementation
  * bug 460 - Rename JWT claims to match spec
  * bug 470 - order does not matter for multi-valued response types
  * bug 471 - Make validateAuthorizeRequest available for POST in addition to GET
  * bug 475 - Adds JTI table definitiion
  * bug 481 - better randomness for generating access tokens
  * bug 480 - Use hash_equals() for signature verification (prevents remote timing attacks)
  * bugs 489, 491, 498 - misc other fixes

* 1.5 (2014-08-27)

  PR: https://github.com/bshaffer/oauth2-server-php/pull/446

  * bug #399 - Add DynamoDB Support
  * bug #404 - renamed error name for malformed/expired tokens
  * bug #412 - Openid connect: fixes for claims with more than one scope / Add support for the prompt parameter ('consent' and 'none')
  * bug #411 - fixes xml output
  * bug #413 - fixes invalid format error
  * bug #401 - fixes code standards / whitespace
  * bug #354 - bundles PDO SQL with the library
  * [BC] bug #397 - refresh tokens should not be encrypted
  * bug #423 - makes "scope" optional for refresh token storage

* 1.4 (2014-06-12)

  PR: https://github.com/bshaffer/oauth2-server-php/pull/392

  * bug #189 Storage\PDO - allows DSN string in constructor
  * bug #233 Bearer Tokens - allows token in request body for PUT requests
  * bug #346 Fixes open_basedir warning
  * bug #351 Adds OpenID Connect support
  * bug #355 Adds php 5.6 and HHVM to travis.ci testing
  * [BC] bug #358 Adds `getQueryStringIdentifier()` to the GrantType interface
  * bug #363 Encryption\JWT - Allows for subclassing JWT Headers
  * bug #349 Bearer Tokens - adds requestHasToken method for when access tokens are optional
  * bug #301 Encryption\JWT - fixes urlSafeB64Encode(): ensures newlines are replaced as expected
  * bug #323 ResourceController - client_id is no longer required to be returned when calling getAccessToken
  * bug #367 Storage\PDO - adds Postgres support
  * bug #368 Access Tokens - use mcrypt_create_iv or openssl_random_pseudo_bytes to create token string
  * bug #376 Request - allows case insensitive headers
  * bug #384 Storage\PDO - can pass in PDO options in constructor of PDO storage
  * misc fixes #361, #292, #373, #374, #379, #396
* 1.3 (2014-02-27)

  PR: https://github.com/bshaffer/oauth2-server-php/pull/325

  * bug #311 adds cassandra storage
  * bug #298 fixes response code for user credentials grant type
  * bug #318 adds 'use_crypto_tokens' config to Server class for better DX
  * [BC] bug #320 pass client_id to getDefaultScope
  * bug #324 better feedback when running tests
  * bug #335 adds support for non-expiring refresh tokens
  * bug #333 fixes Pdo storage for getClientKey
  * bug #336 fixes Redis storage for expireAuthorizationCode

* 1.2 (2014-01-03)

  PR: https://github.com/bshaffer/oauth2-server-php/pull/288

  * bug #285 changed response header from 200 to 401 when empty token received
  * bug #286 adds documentation and links to spec for not including error messages when no token is supplied
  * bug #280 ensures PHP warnings do not get thrown as a result of an invalid argument to $jwt->decode()
  * bug #279 predis wrong number of arguments
  * bug #277 Securing JS WebApp client secret w/ password grant type

* 1.1 (2013-12-17)

  PR: https://github.com/bshaffer/oauth2-server-php/pull/276

  * bug #278 adds refresh token configuration to Server class
  * bug #274 Supplying a null client_id and client_secret grants API access
  * bug #244 [MongoStorage] More detailed implementation info

src/OAuth2/Response.php
<?php

namespace OAuth2;

/**
 * Class to handle OAuth2 Responses in a graceful way.  Use this interface
 * to output the proper OAuth2 responses.
 *
 * @see OAuth2\ResponseInterface
 *
 * This class borrows heavily from the Symfony2 Framework and is part of the symfony package
 * @see Symfony\Component\HttpFoundation\Request (https://github.com/symfony/symfony)
 */
class Response implements ResponseInterface
{
    public $version;
    protected $statusCode = 200;
    protected $statusText;
    protected $parameters = array();
    protected $httpHeaders = array();

    public static $statusTexts = array(
        100 => 'Continue',
        101 => 'Switching Protocols',
        200 => 'OK',
        201 => 'Created',
        202 => 'Accepted',
        203 => 'Non-Authoritative Information',
        204 => 'No Content',
        205 => 'Reset Content',
        206 => 'Partial Content',
        300 => 'Multiple Choices',
        301 => 'Moved Permanently',
        302 => 'Found',
        303 => 'See Other',
        304 => 'Not Modified',
        305 => 'Use Proxy',
        307 => 'Temporary Redirect',
        400 => 'Bad Request',
        401 => 'Unauthorized',
        402 => 'Payment Required',
        403 => 'Forbidden',
        404 => 'Not Found',
        405 => 'Method Not Allowed',
        406 => 'Not Acceptable',
        407 => 'Proxy Authentication Required',
        408 => 'Request Timeout',
        409 => 'Conflict',
        410 => 'Gone',
        411 => 'Length Required',
        412 => 'Precondition Failed',
        413 => 'Request Entity Too Large',
        414 => 'Request-URI Too Long',
        415 => 'Unsupported Media Type',
        416 => 'Requested Range Not Satisfiable',
        417 => 'Expectation Failed',
        418 => 'I\'m a teapot',
        500 => 'Internal Server Error',
        501 => 'Not Implemented',
        502 => 'Bad Gateway',
        503 => 'Service Unavailable',
        504 => 'Gateway Timeout',
        505 => 'HTTP Version Not Supported',
    );

    public function __construct($parameters = array(), $statusCode = 200, $headers = array())
    {
        $this->setParameters($parameters);
        $this->setStatusCode($statusCode);
        $this->setHttpHeaders($headers);
        $this->version = '1.1';
    }

    /**
     * Converts the response object to string containing all headers and the response content.
     *
     * @return string The response with headers and content
     */
    public function __toString()
    {
        $headers = array();
        foreach ($this->httpHeaders as $name => $value) {
            $headers[$name] = (array) $value;
        }

        return
            sprintf('HTTP/%s %s %s', $this->version, $this->statusCode, $this->statusText)."\r\n".
            $this->getHttpHeadersAsString($headers)."\r\n".
            $this->getResponseBody();
    }

    /**
     * Returns the build header line.
     *
     * @param string $name  The header name
     * @param string $value The header value
     *
     * @return string The built header line
     */
    protected function buildHeader($name, $value)
    {
        return sprintf("%s: %s\n", $name, $value);
    }

    public function getStatusCode()
    {
        return $this->statusCode;
    }

    public function setStatusCode($statusCode, $text = null)
    {
        $this->statusCode = (int) $statusCode;
        if ($this->isInvalid()) {
            throw new \InvalidArgumentException(sprintf('The HTTP status code "%s" is not valid.', $statusCode));
        }

        $this->statusText = false === $text ? '' : (null === $text ? self::$statusTexts[$this->statusCode] : $text);
    }

    public function getStatusText()
    {
        return $this->statusText;
    }

    public function getParameters()
    {
        return $this->parameters;
    }

    public function setParameters(array $parameters)
    {
        $this->parameters = $parameters;
    }

    public function addParameters(array $parameters)
    {
        $this->parameters = array_merge($this->parameters, $parameters);
    }

    public function getParameter($name, $default = null)
    {
        return isset($this->parameters[$name]) ? $this->parameters[$name] : $default;
    }

    public function setParameter($name, $value)
    {
        $this->parameters[$name] = $value;
    }

    public function setHttpHeaders(array $httpHeaders)
    {
        $this->httpHeaders = $httpHeaders;
    }

    public function setHttpHeader($name, $value)
    {
        $this->httpHeaders[$name] = $value;
    }

    public function addHttpHeaders(array $httpHeaders)
    {
        $this->httpHeaders = array_merge($this->httpHeaders, $httpHeaders);
    }

    public function getHttpHeaders()
    {
        return $this->httpHeaders;
    }

    public function getHttpHeader($name, $default = null)
    {
        return isset($this->httpHeaders[$name]) ? $this->httpHeaders[$name] : $default;
    }

    public function getResponseBody($format = 'json')
    {
        switch ($format) {
            case 'json':
                return $this->parameters ? json_encode($this->parameters) : '';
            case 'xml':
                // this only works for single-level arrays
                $xml = new \SimpleXMLElement('<response/>');
                foreach ($this->parameters as $key => $param) {
                    $xml->addChild($key, $param);
                }

                return $xml->asXML();
        }

        throw new \InvalidArgumentException(sprintf('The format %s is not supported', $format));

    }

    public function send($format = 'json')
    {
        // headers have already been sent by the developer
        if (headers_sent()) {
            return;
        }

        switch ($format) {
            case 'json':
                $this->setHttpHeader('Content-Type', 'application/json');
                break;
            case 'xml':
                $this->setHttpHeader('Content-Type', 'text/xml');
                break;
        }
        // status
        header(sprintf('HTTP/%s %s %s', $this->version, $this->statusCode, $this->statusText));

        foreach ($this->getHttpHeaders() as $name => $header) {
            header(sprintf('%s: %s', $name, $header));
        }
        echo $this->getResponseBody($format);
    }

    public function setError($statusCode, $error, $errorDescription = null, $errorUri = null)
    {
        $parameters = array(
            'error' => $error,
            'error_description' => $errorDescription,
        );

        if (!is_null($errorUri)) {
            if (strlen($errorUri) > 0 && $errorUri[0] == '#') {
                // we are referencing an oauth bookmark (for brevity)
                $errorUri = 'http://tools.ietf.org/html/rfc6749' . $errorUri;
            }
            $parameters['error_uri'] = $errorUri;
        }

        $httpHeaders = array(
            'Cache-Control' => 'no-store'
        );

        $this->setStatusCode($statusCode);
        $this->addParameters($parameters);
        $this->addHttpHeaders($httpHeaders);

        if (!$this->isClientError() && !$this->isServerError()) {
            throw new \InvalidArgumentException(sprintf('The HTTP status code is not an error ("%s" given).', $statusCode));
        }
    }

    public function setRedirect($statusCode, $url, $state = null, $error = null, $errorDescription = null, $errorUri = null)
    {
        if (empty($url)) {
            throw new \InvalidArgumentException('Cannot redirect to an empty URL.');
        }

        $parameters = array();

        if (!is_null($state)) {
            $parameters['state'] = $state;
        }

        if (!is_null($error)) {
            $this->setError(400, $error, $errorDescription, $errorUri);
        }
        $this->setStatusCode($statusCode);
        $this->addParameters($parameters);

        if (count($this->parameters) > 0) {
            // add parameters to URL redirection
            $parts = parse_url($url);
            $sep = isset($parts['query']) && count($parts['query']) > 0 ? '&' : '?';
            $url .= $sep . http_build_query($this->parameters);
        }

        $this->addHttpHeaders(array('Location' =>  $url));

        if (!$this->isRedirection()) {
            throw new \InvalidArgumentException(sprintf('The HTTP status code is not a redirect ("%s" given).', $statusCode));
        }
    }

    // http://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html
    /**
     * @return Boolean
     *
     * @api
     */
    public function isInvalid()
    {
        return $this->statusCode < 100 || $this->statusCode >= 600;
    }

    /**
     * @return Boolean
     *
     * @api
     */
    public function isInformational()
    {
        return $this->statusCode >= 100 && $this->statusCode < 200;
    }

    /**
     * @return Boolean
     *
     * @api
     */
    public function isSuccessful()
    {
        return $this->statusCode >= 200 && $this->statusCode < 300;
    }

    /**
     * @return Boolean
     *
     * @api
     */
    public function isRedirection()
    {
        return $this->statusCode >= 300 && $this->statusCode < 400;
    }

    /**
     * @return Boolean
     *
     * @api
     */
    public function isClientError()
    {
        return $this->statusCode >= 400 && $this->statusCode < 500;
    }

    /**
     * @return Boolean
     *
     * @api
     */
    public function isServerError()
    {
        return $this->statusCode >= 500 && $this->statusCode < 600;
    }

    /*
     * Functions from Symfony2 HttpFoundation - output pretty header
     */
    private function getHttpHeadersAsString($headers)
    {
        if (count($headers) == 0) {
            return '';
        }

        $max = max(array_map('strlen', array_keys($headers))) + 1;
        $content = '';
        ksort($headers);
        foreach ($headers as $name => $values) {
            foreach ($values as $value) {
                $content .= sprintf("%-{$max}s %s\r\n", $this->beautifyHeaderName($name).':', $value);
            }
        }

        return $content;
    }

    private function beautifyHeaderName($name)
    {
        return preg_replace_callback('/\-(.)/', array($this, 'beautifyCallback'), ucfirst($name));
    }

    private function beautifyCallback($match)
    {
        return '-'.strtoupper($match[1]);
    }
}
